storage/database/db_manager.go

```diff
@@ -1882,9 +1899,87 @@ func (dbm *databaseManager) WritePreimages(number uint64, preimages map[common.H
-	if err := dbm.getDatabase(StateTrieDB).Delete(TrieNodeKey(hash)); err != nil {
-		logger.Crit("Failed to delete trie node", "err", err)
+func (dbm *databaseManager) ReadPruningEnabled() bool {
+	ok, _ := dbm.getDatabase(MiscDB).Has(pruningEnabledKey)
+	return ok
+}
+
+func (dbm *databaseManager) WritePruningEnabled() {
+	if err := dbm.getDatabase(MiscDB).Put(pruningEnabledKey, []byte("42")); err != nil {
+		logger.Crit("Failed to store pruning enabled flag", "err", err)
+	}
+}
+
+func (dbm *databaseManager) DeletePruningEnabled() {
+	if err := dbm.getDatabase(MiscDB).Delete(pruningEnabledKey); err != nil {
+		logger.Crit("Failed to remove pruning enabled flag", "err", err)
+	}
+}
+
+// WritePruningMarks writes the provided set of pruning marks to the database.
+func (dbm *databaseManager) WritePruningMarks(marks []PruningMark) {
+	batch := dbm.NewBatch(MiscDB)
+	for _, mark := range marks {
+		if err := batch.Put(pruningMarkKey(mark), pruningMarkValue); err != nil {
+			logger.Crit("Failed to store trie pruning mark", "err", err)
+		}
+		if _, err := WriteBatchesOverThreshold(batch); err != nil {
+			logger.Crit("Failed to store trie pruning mark", "err", err)
+		}
+	}
+	if err := batch.Write(); err != nil {
+		logger.Crit("Failed to batch write pruning mark", "err", err)
+	}
+}
+
+// ReadPruningMarks reads the pruning marks in the block number range [startNumber, endNumber).
+func (dbm *databaseManager) ReadPruningMarks(startNumber, endNumber uint64) []PruningMark {
+	prefix := pruningMarkPrefix
+	startKey := pruningMarkKey(PruningMark{startNumber, common.ExtHash{}})
+	it := dbm.getDatabase(MiscDB).NewIterator(prefix, startKey[len(prefix):])
+
+	var marks []PruningMark
+	for it.Next() {
+		mark := parsePruningMarkKey(it.Key())
+		if endNumber != 0 && mark.Number >= endNumber {
+			break
+		}
+		marks = append(marks, mark)
+	}
+	return marks
+}
+
+// DeletePruningMarks deletes the provided set of pruning marks from the database.
+// Note that trie nodes are not deleted by this function. To prune trie nodes, use
+// the PruneTrieNodes or DeleteTrieNode functions.
+func (dbm *databaseManager) DeletePruningMarks(marks []PruningMark) {
+	batch := dbm.NewBatch(MiscDB)
+	for _, mark := range marks {
+		if err := batch.Delete(pruningMarkKey(mark)); err != nil {
+			logger.Crit("Failed to delete trie pruning mark", "err", err)
+		}
+		if _, err := WriteBatchesOverThreshold(batch); err != nil {
+			logger.Crit("Failed to delete trie pruning mark", "err", err)
+		}
+	}
+	if err := batch.Write(); err != nil {
+		logger.Crit("Failed to batch delete pruning mark", "err", err)
+	}
+}
+
+// PruneTrieNodes deletes the trie nodes according to the provided set of pruning marks.
+func (dbm *databaseManager) PruneTrieNodes(marks []PruningMark) {
+	batch := dbm.NewBatch(StateTrieDB)
+	for _, mark := range marks {
+		if err := batch.Delete(TrieNodeKey(mark.Hash)); err != nil {
+			logger.Crit("Failed to prune trie node", "err", err)
```

Comment:
Instead of logging and exiting, using the Error level appears to be a more dependable approach. Although the database pruning operation failed for unknown reasons, it is my belief that it did not produce any side effects on other operations, such as block validation, RPC, etc. I would like to know your thoughts on this matter.

Reply:
I agree that PruneTrieNodes failure is not critical. However, it's unlikely that only pruning-related writes fail and others succeed; and making such an exception would degrade the consistency of the database module. Thanks.